tests: boxes: Add tests for saving drafts.

Niloth-p opened the pull request:
### What does this PR do, and why?
Add 2 test functions:
1. Draft message to a channel
2. Draft message to a private narrow (DM and PM huddle)

Test for an existing saved draft for each of the test cases.

Test for:
- a function call to display the save-draft-confirmation popup, in case of existing saved draft
- the created draft object
- return of the key without creating a draft, in case of invalid recipients

This does not test the `_tidy_valid_recipients_and_notify_invalid_ones` or the `session_draft_message` functions. They're mocked. Tests for them would need to be added separately.

### External discussion & connections
- [ ] Discussed in **#zulip-terminal** in `topic`
- [ ] Fully fixes #
- [ ] Partially fixes issue #
- [x] Builds upon previous unmerged work in PR #950 
- [ ] Is a follow-up to work in PR #
- [ ] Requires merge of PR #
- [ ] Merge will enable work on #

### How did you test this?
- [ ] Manually - Behavioral changes
- [ ] Manually - Visual changes
- [ ] Adapting existing automated tests
- [ ] Adding automated tests for new behavior (or missing tests)
- [ ] Existing automated tests should already cover this (*only a refactor of tested code*)

### Self-review checklist for each commit
- [x] It is a [minimal coherent idea](https://github.com/zulip/zulip-terminal#structuring-commits---speeding-up-reviews-merging--development)
- [x] It has a commit summary following the [documented style](https://github.com/zulip/zulip-terminal#structuring-commits---speeding-up-reviews-merging--development) (title & body)
- [ ] It has a commit summary describing the motivation and reasoning for the change
- [x] It individually passes linting and tests
- [ ] It contains test additions for any new behavior
- [ ] It flows clearly from a previous branch commit, and/or prepares for the next commit

Niloth-p committed:
tests: boxes: Add tests for saving drafts.

Test functions added:
- 2 test functions for stream compositions
  - when a draft previously exists, call the confirmation popup
  - when new draft matches saved draft, do not save again
- 3 test functions for private compositions
  - valid recipients,
  - invalid recipients,
  - new draft matches already saved draft

Fixtures added for:
- list saved drafts
- list stream compositions
- list private compositions
- setup for private draft tests
- setup for stream draft tests

Factories added:
- composition factory (generates both stream and private compositions)
- saved draft factory

zormit commented on tests/ui_tools/test_boxes.py:
```diff
@@ -103,6 +105,124 @@ def test_not_calling_typing_method_without_recipients(
+        return_value = write_box.keypress(size, key)
+
+        if not are_valid_recipients:
+            assert key == return_value
```
For my understanding: why are we testing the key here? How is that dependent on the validity of the recipients? It seems to me (from a cursory glance) that `keypress` always returns its `key` argument.

zormit reviewed (commented):
Generally this looks fine to me. I left some comments on things that came to mind. I don't know enough about the codebase yet to give an approval, but given that the test is green and it looks to me like it's testing something useful, I probably would approve it if I had to :P 

hope that helps :)

Niloth-p commented on tests/ui_tools/test_boxes.py:
```diff
@@ -103,6 +105,124 @@ def test_not_calling_typing_method_without_recipients(
+        key: str,
+        mocker: MockerFixture,
+        draft_composition: PrivateComposition,
+        are_valid_recipients: bool,
```
Went with this [Stackoverflow answer](https://stackoverflow.com/a/22397978).
Thanks for the info!

Niloth-p commented on tests/ui_tools/test_boxes.py:
```diff
@@ -103,6 +105,124 @@ def test_not_calling_typing_method_without_recipients(
+        [True, False],
+        ids=["saved_draft_exists", "no_saved_draft_exists"],
+    )
+    def test_keypress_SAVE_DRAFT_MESSAGE_stream(
```
yes, I had pulled that from the unmerged PR, and had failed to notice it and update it. Thanks for the pointer.

neiljp reviewed (commented):
I'll take another look later, but wanted to get this comment out that I'd drafted previously, before mentoring had started.

Niloth-p commented:
Updated to add the `mocked_saved_drafts` list. Took that outside both the test functions to reduce redundancy.
Also switched from using 'huddle' to 'dm_group' in the ids and message.

rsashank commented on tests/ui_tools/test_boxes.py:
```diff
@@ -103,6 +106,144 @@ def test_not_calling_typing_method_without_recipients(
+        write_box.keypress(size, key)
+
+        if not is_every_recipient_valid:
+            assert not write_box.model.save_draft.called
+            assert not write_box.view.controller.save_draft_confirmation_popup.called
```
NIT: Should we prefer using? 
```
write_box.model.save_draft.assert_not_called()
write_box.view.controller.save_draft_confirmation_popup.assert_not_called()
```

neiljp replied:
These are equivalent - or should be!

However, the great thing about using `assert` and the core `called` style values are:
- pytest can dig through levels of code using `assert`, if they fail
- `called` and similar results are values, not assertions in a function

The most useful aspect here though, is that this form can be more easily extracted into parametrize blocks.

rsashank replied:
Hmm, the parametrization part makes sense. Should we stick to using .called from now on?

rsashank commented on tests/ui_tools/test_boxes.py:
```diff
@@ -103,6 +106,144 @@ def test_not_calling_typing_method_without_recipients(
+            assert not write_box.view.controller.save_draft_confirmation_popup.called
+        else:
+            assert write_box.model.session_draft_message.called
+            if draft_saved_in_current_session is not None:
```
NIT: Can this simply be `if draft_saved_in_current_session`?

rsashank replied:
Interesting! Thanks for sharing this :)

My bad, you had to update it again @Niloth-p 😅

rsashank commented:
LGTM! Great work @Niloth-p 👍

Niloth-p commented:
Thank you for catching those details, @rsashank! Great review, that was very helpful.
Updated the requested changes.

neiljp commented on tests/ui_tools/test_boxes.py:
```diff
@@ -103,6 +106,144 @@ def test_not_calling_typing_method_without_recipients(
+        size = widget_size(write_box)
+        write_box.keypress(size, key)
+
+        if not is_every_recipient_valid:
+            write_box.model.save_draft.assert_not_called()
+            write_box.view.controller.save_draft_confirmation_popup.assert_not_called()
+        else:
```
To simplify this, since the test has a nested `if`, I'm wondering if it would be cleaner to extract this first part into a separate test?

That avoids a parametrize over is_every_recipient_valid.

neiljp reviewed (commented):
@Niloth-p I was just mid-review yesterday, so just posting the notes I made and a few more, since I note you updated this already :)

zormit committed:
refactor: Turn saved_draft fixture into dataclass.

zormit committed:
refactor: Make composition_factory produce a Composition.

zormit commented:
@Niloth-p I have tried simplifying and typing your code a bit in https://github.com/zormit/zulip-terminal/tree/1495-save-draft/pr (713bf505, in case I change the branch). What do you think?

I'm still not fully happy with all the layers of indirection, probably it can be simplified even more, but this might give you an idea. I have to stop for now, so I'm sending you this idea, please take it if you like it.

Niloth-p committed:
squash: Make tests pass.

Use different recipients for saved drafts and new drafts, in the
composition factory.

Niloth-p committed:
tests: boxes: Add draft cases for when recipients match saved draft.

Check new drafts with
- same content, different recipients (already exists)
- same recipients, different content (newly added)

Previously, we were using different recipients in the composition
factory, to differentiate the new draft from the saved draft.
Now, it has been extended to allow creating different content as well.

Niloth-p commented:
@zormit Thank you lots for making those changes yourself! That does look much better.

Sorry for the delay in replying back, I'd absorbed your commits right away, but hadn't managed to type out a reply here.

If you find this interesting, please feel free to push your own PR with other improvements and add me as a co-author, that could help get it merged too, as I've done several iterations on this previously, and have been struggling to identify good code and bad code here.

So, no, I'm not really clear on how we can further reduce the indirection, my apologies. I can think of several ways to change the tests, but I have no idea what improves it and what worsens it. If someone is able to guide me with 1 liners like the commit messages, I'll try my best to implement it and get back.

Being new to the complexities of testing, when I come across best practices and guidance, I'm not sure to what extent I should be following them, that I realise I might have overdone it here to the point of complexity, and I'm still not clear enough on all the tradeoffs, so please do let me know how we can proceed from here.

Also, I'm not sure how many commits this is supposed to be. I was initially under the impression this should be a single commit, since it's just tests, and we wouldn't want to be tracking the additions incrementally. But, after the suggestion from @neiljp to break down the refactoring commits of lint-hotkeys, I've been wondering whether we should break this down too. Do we need to?

I've added some newer test cases. But I'm not sure if we want to be that thorough or if it bulks up the testing time unnecessarily.

Niloth-p commented:
Github will report the tests as failed, because I've kept my addition separate from Moritz's commits, but they'll pass if they're squashed. So, please ignore them.
